---
title: "Example_scTypeGSEA"
output: 
  rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{Example_scTypeGSEA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Description

This package is designed to assign cell type labels for each identified cluster in single-cell data. This package uses the “Seurat” R package to do data pre-processing and cell clustering. After clustering the cells, we use differential gene expression analysis to compute the fold-change in gene expression by comparing the cluster profile against all the other identified clusters together. Then we use the Gene Set Enrichment Analysis (GSEA) technique to compute the enrichment (NES and their associated P-value) of marker genes defined for a set of cell types. GSEA analysis provides us the most statistically relevant cell type for each cluster that is finally assigned to the group.

## Installation

You can use following codes to install the package. To use some functions in "fgsea", your R version must be >= 3.6.0.

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("fgsea")
BiocManager::install("MAST")
install.packages("Seurat")

library(devtools)
install_github("cailab-tamu/scTypeGSEA")
```

## Example
```{r, warning=FALSE, message=FALSE}
library(scTypeGSEA)
```

Here we use a small data set "pbmc_small" to show our main function "assignCellType". One can do quality control, data pre-process, cluster, get full changes, do GSEA and label the cell in one step. Following code is for toy data set "pbmc_small".
```{r, warning=FALSE}
pbmc_res <- assignCellType(obj = pbmc_small, min.cells = 1, min.features = 10, nfeatures = 100, 
                           npcs = 10, dims = 1:10, k.param = 5, resolution = 0.75, min.pct = 0.25, 
                           test.use = "MAST", minSize = 5)
```

It will return a list with 2 slots. The first slot is a Seurat object.
```{r}
pbmc_res$obj
```

The second slot is a matrix contains cluster-ID, cell type and p-value for each cell.
```{r}
head(pbmc_res$cell_mat)
```

Next is to illustrate this pipeline step by step.

### Load data and quality control
Here we will use data set "pbmc" go through this pipline. First is to load data set and do quality control. The input can be a Seurat object or count gene expression matrix.
```{r, warning=FALSE}
library(scTypeGSEA)
pbmc <- scqc(pbmc_raw, min.cells = 3, min.features = 200, percent.mt = 5)
pbmc
```

### Data pre-process and do cluster
Next to do data pre-process and cluster. Here data pre-process includes normalization, feature selection and dimension reduction. Here we do “LogNormalize” nomalization, "vst" feature selections to select 2000 features and do PCA to get first 50 PCs. After data-process, we cluster the data.
```{r, warning=FALSE}
set.seed(47)
pbmc <- doClustering(pbmc, normalization.method = "LogNormalize", scale.factor = 10000, 
                     selection.method = "vst", nfeatures = 2000, npcs = 50, dims = 1:50, 
                     k.param = 20, resolution = 0.5)
head(pbmc@meta.data$seurat_clusters)
```

If your data set have already been clustered, you can set "doit = TRUE" to do cluster using new parameters.
```{r, eval=FALSE}
## don't run
pbmc <- check_cluster(pbmc, k.param = 30, resolution = 0.5, doit = TRUE)
```

### Get full changes
To get rank of gene list, we need to do gene differential expression analysis for each cluster with respect to all others. Here we use "wilcox" method.
```{r, warning=FALSE}
cluster_list <- getFC(pbmc, min.pct = 0.25, test.use = "wilcox")
head(cluster_list[[1]])
```

### Gene set enrichment analysis (GSEA)
After getting rank of gene list for each cluster, GSEA method will give us NSF. By NSF, we can decide the cell type for each cluster. Here we use "PanglaoDB" database.
```{r, warning=FALSE}
cluster_celltype <- doGSEA(cluster_list = cluster_list, db = "PanglaoDB_list",
                                  minSize = 15, maxSize = 500)
head(cluster_celltype)
```

You can also use your own cell type database. Here "otherdb" can be a path to the new data base that hope to be used, which is list of cell types with their marker genes. The file must be 'rds' format.
```{r, eval=FALSE}
## Don't run
cluster_celltype <- GSEA_analysis(cluster_list = cluster_list, otherdb = "path/to/your/rdsfile")
```

### Label the cell type
The last step is to add label cell type to our Seurat object.
```{r, warning=FALSE}
pbmc_res <- labelSeurat(pbmc, cluster_celltype)
pbmc_res$obj
head(pbmc_res$cell_mat)
```

### Plot the data with cell type

```{r, warning=FALSE}
library(Seurat)
pbmc <- pbmc_res$obj
pbmc <- Seurat::RunTSNE(pbmc, dims = 1:30)
Seurat::DimPlot(pbmc, reduction = "tsne", label = TRUE, pt.size = 0.5) + NoLegend()
```
